src/ContainerSetPathCapableTrait.php

```diff
@@ -0,0 +1,95 @@
+<?php
+
+namespace Dhii\Data\Container;
+
+use Dhii\Util\String\StringableInterface as Stringable;
+use Psr\Container\ContainerInterface as BaseContainerInterface;
+
+/**
+ * Functionality for getting data from nested container.
+ *
+ * @since [*next-version*]
+ */
+trait ContainerSetPathCapableTrait
+{
+    /**
+     * Sets data on the nested container.
+     *
+     * @since [*next-version*]
+     *
+     * @param array|ArrayAccess|stdClass|BaseContainerInterface $container The top container in the chain.
+     * @param array|Traversable|stdClass $path The list of path segments.
+     * @param mixed                      $value The value to set by path.
+     *
+     * @throws ContainerExceptionInterface If an error occurred while reading or writing from one of the containers in the chain.
+     * @throws InvalidArgumentException If one of the containers in the chain is invalid.
+     * @throws NotFoundExceptionInterface If one of the containers in the chain does not have the corresponding key.
+     * @throws OutOfRangeException
+     */
+    protected function _containerSetPath(&$container, $path, $value)
+    {
+        $path = $this->_normalizeArray($path);
+
+        if (!count($path)) {
+            return;
+        }
+
+        if (count($path) === 1) {
+            $this->_containerSet($container, $path[0], $value);
+            return;
+        }
+
+        $popped = array_shift($path);
```

Comment:
A shifted piece is called `$popped`... I would say, something more semantic, like `$currentSegment`.